### DISCUSSION TITLE
Cannot set “use” : “enc” with custom ECDSA key Provider

### DISCUSSION THREAD

**Discussion by edmaputra** [keycloak/keycloak#8662] [Q&A]
Hi,
I have implemented a custom SPI for Key Provider to create a Key with EC, combine with nimbus-jose-jwt library so I can create a key with ECDH-ES+A128KW algorithm.

The key is generated successfully, but when I want to publish it in oidc jwks uri, the “use” keeps on “sig” value, even I set it to the “enc” in the configuration. I also try to force the “use” value to “enc” in the code but it still happening. Can someone explain why and what should I do ?

![image](https://user-images.githubusercontent.com/2734569/139307939-c52452a8-a3bf-4339-b7a8-1f5851ef4287.png)

![image](https://user-images.githubusercontent.com/2734569/139307963-47e5d6a7-533b-438c-acca-47817a26c423.png)

![image](https://user-images.githubusercontent.com/2734569/139308602-a69db8c7-aba7-4ac1-8065-bf9592556b12.png)

**Comment by tnorimat** [comment]
Hello @edmaputra ,

I'm not sure which Key Provider you've implemented and how, but I assume that you've extended `AbstractEcdsaKeyProvider` and `AbstractEcdsaKeyProviderFactory` or `GeneratedEcdsaKeyProviderFactory` and `GeneratedEcdsaKeyProvider` .

The issue you've reported might be occurred due to the following code.

https://github.com/keycloak/keycloak/blob/9dfcaf016291c17894547fd1bbbcd684621ef605/services/src/main/java/org/keycloak/keys/AbstractEcdsaKeyProvider.java#L64

I'm not sure it works well but how about overriding this `createKeyWrapper` method? 
E.g.
``` java
             KeyUse keyUse = KeyUse.valueOf(model.get(Attributes.KEY_USE, KeyUse.SIG.name()).toUpperCase());
             key.setUse(keyUse == null ? KeyUse.SIG : keyUse);
```

**Reply by edmaputra** [reply]
Hi @tnorimat 

Yes, I already use the Components you've mention before. I also override the `createKeyWrapper` method.

I have found the cause. 

https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java#L125

Invoked by this method to publish the JWKS 
https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java#L239

So whatever the KeyUse I set, it will always publish as "sig" value.

My current solution now is implementing a new custom REST API to publish the key.

**Reply by tnorimat** [reply]
Hello @edmaputra ,

Yes, as you mentioned, the current keycloak does not consider EC key provider for key encryption.
This matter cannot be resolved by only modifying key provider. Therefore, it might be needed to modify keycloak's body codes.
